contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,244 @@
+    }
+    _;
+  }
+
+  /**
+   * @notice Register a new rewards entity
+   * @param entity The entity address to register
+   * @param requiresApproval Whether the entity requires approval for agreements
+   */
+  function registerRewardsEntity(
+    address entity,
+    bool requiresApproval
+  ) external {
+    if (entity == address(0)) {
+      revert InvalidEntityAddress(entity);
+    }
+
+    if (_entities[entity]) {
+      revert EntityAlreadyExists(entity);
+    }
+
+    _entities[entity] = true;
+    _requiresApproval[entity] = requiresApproval;
+    emit RewardsEntityRegistered(entity, requiresApproval);
+  }
+
+  /**
+   * @notice Set whether a Rewards Entity requires approval for agreements
+   * @param requiresApproval Whether the entity requires approval
+   */
+  function setRequiresApprovalForRewardsAgreements(
+    bool requiresApproval
+  ) external entityExists(msg.sender) {
+    _requiresApproval[msg.sender] = requiresApproval;
+    emit RequiresApprovalForRewardsAgreements(msg.sender, requiresApproval);
+  }
+
+  /**
+   * @notice Register a Rewards Consumer - Rewards Provider relationship
+   * @dev Should be called by the Rewards Consumer
+   * @param rewardsProvider The provider entity address
+   */
+  function registerRewardsAgreement(
+    address rewardsProvider
+  ) external entityExists(rewardsProvider) entityExists(msg.sender) {
+    // If the provider requires approval, revert the transaction
+    if (_requiresApproval[rewardsProvider]) {
+      revert ProviderRequiresApproval(rewardsProvider);
+    }
+
+    // Check if agreement already exists
+    bytes32 agreementKey = keccak256(
+      abi.encodePacked(rewardsProvider, msg.sender)
+    );
+    if (_agreements[agreementKey]) {
+      revert AgreementAlreadyExists(rewardsProvider, msg.sender);
+    }
+
+    _agreements[agreementKey] = true;
+    emit RewardsAgreementRegistered(rewardsProvider, msg.sender);
+  }
+
+  /**
+   * @notice Approve a rewards agreement
+   * @dev Should be called by the Rewards Provider
+   * @param rewardsConsumer The consumer entity address
+   */
+  function approveRewardsAgreement(
+    address rewardsConsumer
+  ) external entityExists(rewardsConsumer) entityExists(msg.sender) {
+    // Create the agreement
+    bytes32 agreementKey = keccak256(
+      abi.encodePacked(msg.sender, rewardsConsumer)
+    );
+    _agreements[agreementKey] = true;
+    emit RewardsAgreementApproved(msg.sender, rewardsConsumer);
+  }
+
+  /**
+   * @notice Register a user as being referred to a rewards agreement
+   * @dev Requires REFERRAL_REGISTRAR_ROLE
+   * @param user The address of the user being referred
+   * @param rewardsConsumer The address of the rewards consumer entity
+   * @param rewardsProvider The address of the rewards provider entity
+   */
+  function registerReferral(
+    address user,
+    address rewardsConsumer,
+    address rewardsProvider
+  ) external entityExists(rewardsProvider) entityExists(rewardsConsumer) {
```

Comment:
Can we use the OpenZeppelin `onlyRole` modifier here? (see [here](https://docs.openzeppelin.com/contracts/4.x/access-control#using-access-control)) Or are we not using it explicitly so that we can emit a custom revert reason?